distributed/event.py

```diff
@@ -0,0 +1,251 @@
+        self._events = dict()
+        # Keep track on how many waiters are present, so we know when
+        # we can remove the event
+        self._waiter_count = defaultdict(int)
+
+        self.scheduler.handlers.update(
+            {
+                "event_wait": self.event_wait,
+                "event_set": self.event_set,
+                "event_clear": self.event_clear,
+                "event_is_set": self.event_is_set,
+            }
+        )
+
+        self.scheduler.extensions["events"] = self
+
+    async def event_wait(self, stream=None, name=None, timeout=None):
+        """ Wait until the event is set to true.
+        Returns false, when this did not happen in the given time
+        and true otherwise.
+        """
+        with log_errors():
+            name = self._normalize_name(name)
+
+            event = self._get_or_create_event(name)
+            future = event.wait()
+            if timeout is not None:
+                future = asyncio.wait_for(future, timeout)
+
+            self._waiter_count[name] += 1
+            try:
+                await future
+            except TimeoutError:
+                return False
+            finally:
+                self._waiter_count[name] -= 1
+
+                if not self._waiter_count[name] and not event.is_set():
+                    # No one is waiting for this
+                    # and as the default flag for an event is false
+                    # we can safely remove it
+                    self._delete_event(name)
+
+            return True
+
+    async def event_set(self, stream=None, name=None):
+        """ Set the event with the given name to true.
+
+        All waiters on this event will be notified.
+        """
+        with log_errors():
+            name = self._normalize_name(name)
+            # No matter if someone is listening or not,
+            # we set the event to true
+            event = self._get_or_create_event(name)
+            self.scheduler.loop.add_callback(event.set)
+
+    async def event_clear(self, stream=None, name=None):
+        """Set the event with the given name to false."""
+        with log_errors():
+            name = self._normalize_name(name)
+            if not self._waiter_count[name]:
+                # No one is waiting for this
+                # and as the default flag for an event is false
+                # we can safely remove it
+                self._delete_event(name)
+
+            else:
+                # There are waiters
+                # This can happen if an event is "double-cleared"
+                # In principle, the event should be unset at this point
+                # (because if it is set, all waiters should have been
+                # notified). But to prevent race conditions
+                # do to unlucky timing, we clear anyways
+                assert name in self._events
+                event = self._events[name]
+                self.scheduler.loop.add_callback(event.clear)
```

Comment:
If possible I would prefer running `event.clear` immediately.  Running it later is the sort of thing that can lead to race conditions.  Right now we know that we can safely clear it so great, let's clear it now rather than later.